gh-103921: Document PEP 695 (#104642)
python/cpython: Doc/whatsnew/3.12.rst

(PEP written by Franek Magiera)

.. _whatsnew312-pep695:

PEP 695: Type Parameter Syntax
------------------------------

Generic classes and functions under :pep:`484` were declared using a verbose syntax
that left the scope of type parameters unclear and required explicit declarations of
variance.

:pep:`695` introduces a new, more compact and explicit way to create
:ref:`generic classes <generic-classes>` and :ref:`functions <generic-functions>`::

   def max[T](args: Iterable[T]) -> T:
       ...

   class list[T]:
       def __getitem__(self, index: int, /) -> T:
           ...

       def append(self, element: T) -> None:
           ...

> In addition, the PEP introduces a new way to write :ref:`type aliases <type-aliases>`

Reviewer: ```suggestion
In addition, the PEP introduces a new way to declare :ref:`type aliases <type-aliases>`
```

using the :keyword:`type` statement, which creates an instance of
:class:`~typing.TypeAliasType`::

   type Point = tuple[float, float]

Type aliases can also be :ref:`generic <generic-type-aliases>`::

   type Point[T] = tuple[T, T]

> The new syntax also allows declaring :class:`~typing.TypeVarTuple`

Reviewer: The previous paragraph also starts with "also"

```suggestion
The new syntax allows declaring :class:`~typing.TypeVarTuple`
```

and :class:`~typing.ParamSpec` parameters, as well as :class:`~typing.TypeVar`
parameters with bounds or constraints::

   type IntFunc[**P] = Callable[P, int]  # ParamSpec
   type LabeledTuple[*Ts] = tuple[str, *Ts]  # TypeVarTuple
   type HashableSequence[T: Hashable] = Sequence[T]  # TypeVar with bound
   type IntOrStrSequence[T: (int, str)] = Sequence[T]  # TypeVar with constraints

> The value of type aliases and the bound and constraints of type variables
> created through this syntax are evaluated only on demand, enabling the use of
> forward references to types defined later in the file.

Reviewer: Shorter sentences:

```suggestion
The value of type aliases and the bound and constraints of type variables
created through this syntax are evaluated only on demand. This means type
aliases are able to refer to other types defined later in the file.
```

Also, could you link to the section on lazy evaluation here?

Author: Done!

Type parameters declared through a type parameter list are visible within the
scope of the declaration and any nested scopes, but not in the outer scope. For
example, they can be used in the type annotations for the methods of a generic
class or in the class body. However, they cannot be used in the module scope after

> the class is defined.

Reviewer: This is quite technical, and hard to understand without an example; I wonder if it belongs in whatsnew. Maybe we should just link to some more in-depth documentation elsewhere?

Author: I'll add a reference to :ref:`type-params`. I think it's important to mention this to ensure people know this is not just a new typing feature but also a change to a pretty core part of the language.
